enumerable_project.rb

```diff
@@ -0,0 +1,387 @@
+# rubocop:disable Metrics/CyclomaticComplexity
+
+# rubocop:disable Metrics/PerceivedComplexity
+
+# rubocop:disable Metrics/ModuleLength
+
+module Enumerable
+  def my_each
+    if self.class == Array
+
+      0.upto(self.length - 1) do |i|
+        yield(self[i])
+      end
+
+    elsif self.class == Hash
+
+      for key in self
+
+        yield(key[0].to_s, key[1])
+
+      end
+
+    end
+  end
+
+  def my_each_with_index
+    if self.class == Array
+
+      0.upto(self.length - 1) do |i|
+        yield(self[i], i)
+      end
+
+    elsif self.class == Hash
+
+      i = 0
+
+      for key in self
+
+        yield([key[0].to_s, key[1]], i)
+
+        i+=1
+
+      end
+
+    end
+  end
+
+  def my_select
+    if self.class == Array
+
+      arr = []
+
+      0.upto(self.length-1) do |i|
+        arr << self[i] if yield(self[i])
+      end
+
+      return arr
+
+    elsif self.class == Hash
+
+      hash = {}
+
+      for key in self
+
+        hash[key[0]] = key[1] if yield(key[0], key[1])
+
+      end
+
+      return hash
+
+    end
+  end
+
+  def my_all?
+    if self.class == Array
+
+      0.upto(self.length-1) do |i|
+        return false unless yield(self[i])
+      end
+
+      return true
+
+    elsif self.class == Hash
+
+      for key in self
+
+        return false unless yield(key[0], key[1])
+
+      end
+
+      return true
+
+    end
+  end
+
+  def my_none?
+    if self.class == Array
+
+      0.upto(self.length-1) do |i|
+        return false if yield(self[i])
+      end
+
+      return true
+
+    elsif self.class == Hash
+
+      for key in self
+
+        return false if yield(key[0], key[1])
+
+      end
+
+      return true
+
+    end
+  end
+
+  def my_count
+    count = 0
+
+    if self.class == Array
+
+      0.upto(self.length-1) do |i|
+        count += 1 if yield(self[i])
+      end
+
+    elsif self.class == Hash
+
+      for key in self
+
+        count += 1 if yield(key[0], key[1])
+
+      end
+
+    end
+
+    return count
+  end
+
+  def my_map(my_proc = nil)
+    arr = []
+
+    if self.class == Array
+
+      0.upto(self.length - 1) do |i|
+        arr << (my_proc ? my_proc.call(i) : yield(self[i]))
+      end
+
+    elsif self.class == Hash
+
+      for key in self
+
+        arr << (my_proc ? my_proc.call(key[0], key[1]) : yield(key[0], key[1]))
+
+      end
+
+    end
+
+    return arr
+  end
```

Comment:
* [ ] should have working `my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_map 

- [ ] returns an Enumerator if no block is given
```ruby
[2,4,7,11].my_map should be equal to [2,4,7,11].map #<Enumerator: [2, 4, 7, 11]:map
```
- Yours is always expecting for a block
![image](https://user-images.githubusercontent.com/52209067/78953617-b9da5c80-7a96-11ea-9978-0fc69e5fb8d6.png)

- Real map
![image](https://user-images.githubusercontent.com/52209067/77697646-dc3b8880-6f74-11ea-9277-4ed4a2b3af4c.png)

- [ ] in the specifications of the project in the [Odin project](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks) # 12, you were asked to modify your my_map method to take either a proc or a block. And if both a proc and a block are given, only execute the proc. The real map doesn't have this feature, but you were asked to have it so I strongly recommend working on it

Given 
```ruby
my_proc = Proc.new {|num| num > 10 }
```
Result
```ruby
puts [18, 22, 5, 6] .my_map(my_proc) {|num| num < 10 }  => true true false false
```